### PR TITLE
Tests: add regrid corner case tests

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Add tests for corner-cases for the regrid method. Many of these tests are currently regression tests, and check that the code behaviour has not changed, rather than that it works correctly. There is no functional change to the code.

# Details

These tests have been lying around for a while but I am not convinced the PRs they come from are going to make it into 4.16, so I want the actual tests in so I don't keep on re-inventing them, and also so we can easily see when the behaviour changes (hopefully fixed).

- adds explicit tests of #1802 
- taken from 
  - #1179 
  - #1803 
 although much of this code has been re-worked from previous un-merged PRs

### CODE REVIEW

File: sherpa/astro/xspec/tests/test_xspec_regrid.py [hunk 1]
```diff
@@ -0,0 +1,639 @@
...
+    mdl2 = xsmodel(name2)
+    cmdl = conv(mdl1 * mdl2)
+
+    rmdl = cmdl.regrid(ebase[:-1], ebase[1:])
+
+    exp = cmdl(eg1, eg2)
+    assert (exp > 0).all()
+
+    y = rmdl(eg1, eg2)
+    assert y == pytest.approx(exp, rel=0.35)
```
Comment (hamogu) [inline sherpa/astro/xspec/tests/test_xspec_regrid.py:639]: It's so large that I wonder if something is wrong here... but as long as essentially everything is expected to fail anyway, I won't worry.

**Review by hamogu — APPROVED** [review]

File: sherpa/astro/xspec/tests/test_xspec_regrid.py (hunk 1 above)
Comment (DougBurke) [reply sherpa/astro/xspec/tests/test_xspec_regrid.py:639]: I am treating this as a regression test at the moment, but it is one of the things I have been trying to get to the bottom of with my work on the regrid code, but have not been able to come to a satisfactory story just yet.
